### ISSUE TITLE
CURS does not properly handle sorting versions

### ISSUE THREAD

**Issue by ugexe** [rakudo/rakudo#4843]
When a CUR is queried for a module it takes the highest version from the first repository with any version. So if a CURI has a `Baz:ver<100>`, and a CURS for that same CURI contains a `Baz:ver<1>` (for say a dependency of `Bar`, which `Foo` depends on), then a test for `Foo` that has `use Baz:ver<1+>` will use version `1` instead of `100`.

To fix this CURS needs to implement its own `candidates` method that looks at distributions in its own repository (those installed to the CURS) as well as those of the CURI it is shadowing. However this also means that when the CURS does not find a given module that the subsequent search on its CURI will also fail (ideally we wouldn't make this second query at all since we know it will fail).

**Comment by lizmat** [comment]
If I get this correctly, if both CURS and CURI return candidates, then we only need to check for the highest version, right?

**Comment by lizmat** [comment]
A first stab:
````raku
    method candidates(|c) {
        if callsame() -> @staging_candidates {
            if $!parent.candidates(|c) -> @installed_candidates {
                (|@installed_candidates, |@staging_candidates)
                  .sort(*.meta<ver>)
                  .reverse
            }
            else {
                @staging_candidates
            }
        }
        else {
            $!parent.candidates(|c)
        }
    }
````
Comments ?  Suggestions?

**Comment by lizmat** [comment]
> However this also means that when the CURS does not find a given module that the subsequent search on its CURI will also fail (ideally we wouldn't make this second query at all since we know it will fail).

I'm not grokking the reasoning behind this.  Surely if the CURS doesn't have a module matching the dependency spec, surely the CURI one should be returned if available?  How else would `use Test` work on a CURS on the `core` repository?

**Comment by ugexe** [comment]
> I'm not grokking the reasoning behind this. Surely if the CURS doesn't have a module matching the dependency spec, surely the CURI one should be returned if available? How else would use Test work on a CURS on the core repository?

CURS should emulate having all the modules available of whatever CURI it is shadowing. Therefore if `Test` isn't available for a given CURS then it must not be available from the specific CURI it is shadowing.

**Comment by ugexe** [comment]
> Comments ? Suggestions?

Its not considering sorting of the versions. You cannot simply select candidates from one or the other, you must get the candidates from both and then sort.

**Comment by ugexe** [comment]
> Comments ? Suggestions?

I think that is the gist of it other than handling `api` sorting.

**Comment by lizmat** [comment]
But if the CURS doesn't have any, then it should just return the CURI ones?  And if the CURS has candidates, and the CURI doesn't, then it should just return the CURS ones?  Only if they have both, should we merge and sort, no?

**Comment by lizmat** [comment]
Added `api` sorting.  Having said that, I still think we should disallow distributions with the same `ver` but a different `api` (which would make sorting on `api` not necessary).

**Comment by ugexe** [comment]
> Added api sorting. Having said that, I still think we should disallow distributions with the same ver but a different api (which would make sorting on api not necessary).

I think the current api handling is fine, which I've pondered on for nearly five years now.

**Comment by ugexe** [comment]
> (|@installed_candidates, |@staging_candidates)

Technically I *think* this should be reversed. The ordering of matches when the `api` and `ver` match (and just the `auth` differs) is based on whichever one is installed first (i.e. what is written first to the short-id lookup of a given CURI).

**Comment by lizmat** [comment]
Reverted the order.  Which would mean that if the CURS has a candidate with the same `ver` and `api` as a candidate in `CURI`, it would select the CURS one due to stable sorting.  That's what you mean, right?

**Comment by lizmat** [comment]
Oops.

**Comment by ugexe** [comment]
> Reverted the order. Which would mean that if the CURS has a candidate with the same ver and api as a candidate in CURI, it would select the CURS one due to stable sorting. That's what you mean, right?

The CURI should be chosen (since when the CURS files are copied over the one originally installed in the CURI would be first in the short-id lookup). Originally it had `(|@installed_candidates, |@staging_candidates)` which when reversed would (I think) incorrectly put the staging candidate in front.

**Comment by lizmat** [comment]
> the desired behavior would be for the CURI candidate to be chosen given a dist in a shadowing CURS with the same name, ver, and api (only differing in auth)

That feels... weird.  Suppose one is working on a copy of the one in CURI, but under your own `auth`: surely you'd want the CURI one to be selected, e.g. for testing?

**Comment by lizmat** [comment]
ok, reverted the order (again).

**Comment by ugexe** [comment]
So I was thinking more about the query order and it occured to me (and maybe you alluded to earlier and I overlooked) that this would be a bit closer:

```raku
    method candidates(|c) {
    	# If the repo we are shadowing isn't otherwise the last in the chain then
    	# make sure we don't inadver
	for $*REPO.repo-chain.grep(CompUnit::Repository::Installation) -> $curi {
	    last if $curi ~~ $self;
	    return $_ with $curi.candidates(|c)
	}

        if callsame() -> @staging_candidates {
            if $!parent.candidates(|c) -> @installed_candidates {
                (|@installed_candidates, |@staging_candidates)
                  .sort(*.meta<ver>)
                  .sort(*.meta<api>)
                  .reverse
            }
            else {
                @staging_candidates
            }
        }
        else {
            $!parent.candidates(|c)
        }
    }
```
...which tries to ensure that repos that come before the one being shadowed do not get shadowed by the unrelated CURS.

Thats still not quite perfect either though, since it doesn't respect non-`CompUnit::Repository::Installation` repos. `for $*REPO.repo-chain.grep(*.^can("candidates"))` might be closer yet, as it would respect all raku distribution `CompUnit::Repository`.

An (ideal) alternative to the theoretical  changes I suggested earlier in this reply is figuring out if we can somehow automatically move any `CURS` to be directly in front of the `CURI` it shadows in the repo chain. I'm not sure if this would have any negative implications though.

**Comment by lizmat** [comment]
But doesn't the CURS already **replace** the CURI in the chain with this:
````raku
CompUnit::RepositoryRegistry.register-name($!name, self);
````
and thus doesn't the CURS already **just** shadow the CURI it is staging??

**Comment by niner** [comment]
On Samstag, 2. April 2022 00:21:05 CEST Elizabeth Mattijsen wrote:
> But doesn't the CURS already **replace** the CURI in the chain with this:

No, that just replaces the CURI in the by-name lookup hash. It does not touch 
the $*REPO chain at all.

**Comment by lizmat** [comment]
> It does not touch the $*REPO chain at all.

So maybe it should also replace it in the `$*REPO` chain as well?  That could be done in `TWEAK`, and then have `self-destruct` restore the original situation?

**Comment by niner** [comment]
> So maybe it should also replace it in the `$*REPO` chain as well? That
> could be done in `TWEAK`, and then have `self-destruct` restore the
> original situation?

Why should it? CURS is a very fringe feature. You can expect its users to be 
somewhat knowledgeable about module loading and repository chains. And the 
number of users is very limited, so we're not saving humanity a lot of time.

Doing this unconditionally in TWEAK would lock us into a certain usage 
pattern. We don't know what other patterns may become interesting in the 
future.

**Comment by lizmat** [comment]
> Doing this unconditionally in TWEAK would lock us into a certain usage pattern.

But isn't that also the case with `CompUnit::RepositoryRegistry.register-name($!name, self);`

I'd suggest we add a `:shadow-completely` parameter to have it *both* re-register with the repo name, *and* replace it in the `$*REPO` chain.

**Comment by lizmat** [comment]
> CURS is a very fringe feature.

I disagree.  I think it's one of the more important features of using Raku in production in larger business environments.  It allows a company to set up its own sysadmin flows for Raku distribution installation.
